oci-image/src/state.rs

```diff
@@ -0,0 +1,216 @@
+    pub digest: String,
+    /// The image layers metas
+    pub layers: Vec<LayerMetadata>,
+    /// The image configuration
+    pub config: ImageConfiguration,
+}
+
+/// `State` is responsible about storing information's about images and layers.
+#[derive(Clone, Default, Deserialize, Debug, Serialize, PartialEq)]
+pub(crate) struct State {
+    /// The path to the state file
+    #[serde(skip)]
+    pub state_file: PathBuf,
+    /// An hashmap that holds every images pulled
+    images: HashMap<String, ImageMetadata>,
+    /// An hashmap that holds every layers pulled
+    layers: HashMap<String, LayerMetadata>,
+    /// An hashmap that holds snapshots with working directory index
+    snapshots: HashMap<String, usize>,
+}
+
+impl State {
+    /// Save the state into the file.
+    pub fn save(&self) -> crate::Result<()> {
+        let serialized = serde_json::to_string_pretty(&self)
+            .map_err(|e| Error::SerializeState(e.to_string()))?;
+
+        OpenOptions::new()
+            .write(true)
+            .open(&self.state_file)
+            .map_err(|e| Error::OpenStateFile(e.to_string()))?
+            .write_all(serialized.as_bytes())
+            .map_err(|e| Error::WriteStateFile(e.to_string()))
+    }
+}
+
+impl MetadataManager for State {
+    fn add_image(&mut self, image: &ImageMetadata) -> &mut State {
+        self.images.insert(image.id.clone(), image.clone());
+        self
+    }
+
+    fn add_layer(&mut self, layer: &LayerMetadata) -> &mut State {
+        self.layers
+            .insert(layer.compressed_digest.clone(), layer.clone());
+        self
+    }
+
+    fn add_snapshot(&mut self, index: usize) -> &mut Self {
+        self.snapshots.insert("overlay".to_string(), index);
+        self
+    }
+
+    fn has_image(&self, image_id: &str) -> bool {
+        self.images.contains_key(image_id)
+    }
+
+    fn has_layer(&self, layer_id: &str) -> bool {
+        self.layers.contains_key(layer_id)
+    }
+
+    fn has_snapshot(&self, snapshot_id: &str) -> bool {
+        self.layers.contains_key(snapshot_id)
+    }
+
+    fn get_image(&self, image_id: &str) -> Option<&ImageMetadata> {
+        self.images.get(image_id)
+    }
+
+    fn get_layer(&self, layer_id: &str) -> Option<&LayerMetadata> {
+        self.layers.get(layer_id)
+    }
+
+    fn get_snapshot(&self, snapshot_id: &str) -> Option<&usize> {
+        self.snapshots.get(snapshot_id)
+    }
+}
+
+impl TryFrom<&PathBuf> for State {
+    type Error = crate::Error;
+
+    fn try_from(state_file: &PathBuf) -> Result<Self, Self::Error> {
+        let file = match state_file.exists() {
+            true => File::open(state_file).map_err(|e| Error::OpenStateFile(e.to_string()))?,
+            false => File::create(state_file).map_err(|e| Error::OpenStateFile(e.to_string()))?,
+        };
+
+        let state = match serde_json::from_reader::<File, State>(file) {
+            Ok(mut state) => {
+                state.state_file = state_file.clone();
```

Comment:
As I dont think you need `state_file`at all, you could just call `map_or`on `serde_json::from_reader::<File, State>(file)`.